src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailedTransportListener.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\EventListener;
+
+use Psr\Container\ContainerInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Debug\Exception\FlattenException;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Messenger\Event\WorkerMessageFailedEvent;
+use Symfony\Component\Messenger\RoutableMessageBus;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+use Symfony\Component\Messenger\Stamp\RedeliveryStamp;
+use Symfony\Component\Messenger\Stamp\SentStamp;
+use Symfony\Component\Messenger\Stamp\SentToFailedTransportStamp;
+
+/**
+ * Sends a rejected message to a "failed transport".
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+class SendFailedMessageToFailedTransportListener implements EventSubscriberInterface
+{
+    private $bus;
+    private $failedSenderAlias;
+    private $logger;
+
+    public function __construct(ContainerInterface $busLocator, string $failedSenderAlias, LoggerInterface $logger = null)
+    {
+        $this->bus = new RoutableMessageBus($busLocator);
+        $this->failedSenderAlias = $failedSenderAlias;
+        $this->logger = $logger;
+    }
+
+    public function onMessageFailed(WorkerMessageFailedEvent $event)
+    {
+        if (true === $event->willRetry()) {
+            return;
+        }
+
+        $envelope = $event->getEnvelope();
+
+        // avoid re-sending to the failed sender
+        foreach ($envelope->all(SentStamp::class) as $sentStamp) {
+            /** @var SentStamp $sentStamp */
+            if ($sentStamp->getSenderAlias() === $this->failedSenderAlias) {
+                return;
+            }
+        }
+
+        // remove the received stamp so it's redelivered
+        $throwable = $event->getThrowable();
+        $flattenedException = \class_exists(FlattenException::class) ? FlattenException::createFromThrowable($throwable) : null;
+        $envelope = $envelope->withoutAll(ReceivedStamp::class)
+            ->with(new SentToFailedTransportStamp($throwable->getMessage(), $event->getReceiverName(), $flattenedException))
+            ->with(new RedeliveryStamp(0, $this->failedSenderAlias));
+
+        if (null !== $this->logger) {
```

Comment:
Why is the logger optional and not required?

Reply:
Because this class is in the component... so no need to make it required. I think that's pretty standard